lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group_metric_filter.py

```diff
@@ -0,0 +1,223 @@
+          required: false
+          type: str
+extends_documentation_fragment:
+    - aws
+    - ec2
+'''
+
+EXAMPLES = '''
+- name: set metric filter on log group /fluentd/testcase
+  cloudwatchlogs_log_group_metric_filter:
+    log_group_name: /fluentd/testcase
+    filter_name: BoxFreeStorage
+    filter_pattern: '{($.value = *) && ($.hostname = "box")}'
+    state: present
+    metric_transformation:
+        metric_name: box_free_space
+        metric_namespace: fluentd_metrics
+        metric_value: "$.value"
+
+- name: delete metric filter on log group /fluentd/testcase
+  cloudwatchlogs_log_group_metric_filter:
+    log_group_name: /fluentd/testcase
+    filter_name: BoxFreeStorage
+    state: absent
+'''
+
+RETURN = """
+metric_filters:
+    description: Return the origin response value
+    returned: success
+    type: list
+    contains:
+        creation_time:
+        filter_name:
+        filter_pattern:
+        log_group_name:
+        metric_filter_count:
+"""
+from ansible.module_utils.aws.core import AnsibleAWSModule, is_boto3_error_code, get_boto3_client_method_parameters
+from ansible.module_utils.ec2 import camel_dict_to_snake_dict
+
+try:
+    from botocore.exceptions import ClientError, BotoCoreError, WaiterError
+except ImportError:
+    pass  # caught by AnsibleAWSModule
+
+
+def metricTransformationHandler(metricTransformations, originMetricTransformations=None):
+
+    if originMetricTransformations:
+        change = False
+        originMetricTransformations = camel_dict_to_snake_dict(
+            originMetricTransformations)
+        for item in ["default_value", "metric_name", "metric_namespace", "metric_value"]:
+            if metricTransformations.get(item) != originMetricTransformations.get(item):
+                change = True
+    else:
+        change = True
+
+    defaultValue = metricTransformations.get("defaultValue")
+    if isinstance(defaultValue, int) or isinstance(defaultValue, float):
+        retval = [
+            {
+                'metricName': metricTransformations.get("metric_name"),
+                'metricNamespace': metricTransformations.get("metric_namespace"),
+                'metricValue': metricTransformations.get("metric_value"),
+                'defaultValue': defaultValue
+            }
+        ]
+    else:
+        retval = [
+            {
+                'metricName': metricTransformations.get("metric_name"),
+                'metricNamespace': metricTransformations.get("metric_namespace"),
+                'metricValue': metricTransformations.get("metric_value"),
+            }
+        ]
+
+    return retval, change
+
+
+def main():
+
+    arg_spec = dict(
+        state=dict(type='str', required=True, choices=['present', 'absent']),
+        log_group_name=dict(type='str', required=True),
+        filter_name=dict(type='str', required=True),
+        filter_pattern=dict(type='str', required=True),
```

Comment:
By always requiring these args, they must be provided for all states. This requires users to enter all details when deleting a filter (and makes the tests to delete metric filters fail). You can do a `required_if` instead like in the cloudwatch module, and update the module docs to say "Required when I(state=present)." . https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/cloudtrail.py#L470

Reply:
Uh, you're right.

```
TASK [cloudwatchlogs : checkmode delete metric filter on 'ansible-test-markus-pc-38389623/integrationtest'] ***
fatal: [testhost]: FAILED! => {"changed": false, "msg": "missing required arguments: filter_pattern, metric_transformation"}

TASK [cloudwatchlogs : delete metric filter] ***********************************
fatal: [testhost]: FAILED! => {"changed": false, "msg": "missing required arguments: filter_pattern, metric_transformation"}
```

fun fact, the integration test told me already about this issue.